### PR TITLE
doc: fix some bugs/warnings in documentation build

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

This PR fixes about 40 various warnings (errors, really) in the docs build.
I also fixed a couple deprecations warnings from matplotlib while I was at it.
At the time of writing there are 37 warnings left to solve, which fall into 3 or 4 distinct categories, but I can't find a solution for any of these at the moment, so I'll open a separate issue for them and consider this PR complete (#3369)

### CODE REVIEW

**Comment by neutrinoceros** [comment]
@yt-fido test this please

File: doc/source/visualizing/callbacks.rst [hunk 1]
```diff
@@ -158,7 +158,7 @@ List Currently Applied Callbacks
 
    This function will print a list of each of the currently applied
    callbacks together with their index.  The index can be used with
-   :ref:`clear_annotations() function <annotate-clear>` to remove a
+   :ref:`clear_annotations() function <clear-annotations>`_ to remove a
```
Comment (neutrinoceros) [inline doc/source/visualizing/callbacks.rst:161]:
```suggestion
   :ref:`clear_annotations function <clear-annotations>`_ to remove a
```

**Comment by neutrinoceros** [comment]
here's the full log of the remaining warnings:
```
WARNING: autodoc: failed to import module 'mesh_construction' from module 'yt.utilities.lib.embree_mesh'; the following exception was raised:
No module named 'yt.utilities.lib.embree_mesh.mesh_construction'
WARNING: autodoc: failed to import module 'mesh_intersection' from module 'yt.utilities.lib.embree_mesh'; the following exception was raised:
No module named 'yt.utilities.lib.embree_mesh.mesh_intersection'
WARNING: autodoc: failed to import module 'mesh_samplers' from module 'yt.utilities.lib.embree_mesh'; the following exception was raised:
No module named 'yt.utilities.lib.embree_mesh.mesh_samplers'
WARNING: autodoc: failed to import module 'mesh_traversal' from module 'yt.utilities.lib.embree_mesh'; the following exception was raised:
No module named 'yt.utilities.lib.embree_mesh.mesh_traversal'
WARNING: autodoc: failed to import module 'convenience' from module 'yt'; the following exception was raised:
Traceback (most recent call last):
  File "/var/jenkins_home/shiningpanda/jobs/c9e5cb26/virtualenvs/d41d8cd9/lib/python3.8/site-packages/sphinx/ext/autodoc/importer.py", line 32, in import_module
    return importlib.import_module(modname)
  File "/var/jenkins_home/shiningpanda/jobs/c9e5cb26/virtualenvs/d41d8cd9/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/jenkins_home/workspace/yt_docs_new/yt/convenience.py", line 3, in <module>
    raise YTModuleRemoved("yt.convenience", "yt.loaders")
yt.utilities.exceptions.YTModuleRemoved: The yt.convenience module has been removed from yt.
It has been moved to yt.loaders.

/var/jenkins_home/workspace/yt_docs_new/yt/data_objects/derived_quantities.py:docstring of yt.data_objects.derived_quantities.AngularMomentumVector:27: WARNING: Inline strong start-string without end-string.
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.data_version:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.data_version, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.domain_width:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.domain_width, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.idx:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.idx, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.leafsize:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.leafsize, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.leaves:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.leaves, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.left_edge:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.left_edge, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.ndim:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.ndim, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.npts:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.npts, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.num_leaves:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.num_leaves, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.periodic:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.periodic, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyKDTree.right_edge:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyKDTree.right_edge, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.domain_width:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.domain_width, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.left_edge:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.left_edge, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.left_neighbors:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.left_neighbors, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.ndim:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.ndim, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.npts:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.npts, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.num_leaves:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.num_leaves, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.periodic_left:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.periodic_left, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.periodic_right:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.periodic_right, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.right_edge:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.right_edge, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.right_neighbors:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.right_neighbors, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.start_idx:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.start_idx, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
docstring of yt.utilities.lib.cykdtree.kdtree.PyNode.stop_idx:1: WARNING: duplicate object description of yt.utilities.lib.cykdtree.kdtree.PyNode.stop_idx, other instance in reference/api/yt.utilities.lib.cykdtree.kdtree, use :noindex: for one of them
/var/jenkins_home/workspace/yt_docs_new/yt/utilities/lib/cykdtree/plot.py:docstring of yt.utilities.lib.cykdtree.plot.plot2D_serial:10: WARNING: Field list ends without a blank line; unexpected unindent.
/var/jenkins_home/workspace/yt_docs_new/yt/visualization/plot_modifications.py:docstring of yt.visualization.plot_modifications.ScaleCallback:59: WARNING: Unexpected indentation.
/var/jenkins_home/workspace/yt_docs_new/yt/visualization/plot_modifications.py:docstring of yt.visualization.plot_modifications.ScaleCallback:56: WARNING: Inline literal start-string without end-string.
/var/jenkins_home/workspace/yt_docs_new/yt/visualization/plot_modifications.py:docstring of yt.visualization.plot_modifications.TimestampCallback:61: WARNING: Unexpected indentation.
/var/jenkins_home/workspace/yt_docs_new/yt/visualization/plot_modifications.py:docstring of yt.visualization.plot_modifications.TimestampCallback:58: WARNING: Inline literal start-string without end-string.
/var/jenkins_home/workspace/yt_docs_new/yt/loaders.py:docstring of yt.loaders.load_sample:28: WARNING: Bullet list ends without a blank line; unexpected unindent.
/var/jenkins_home/workspace/yt_docs_new/yt/loaders.py:docstring of yt.loaders.load_sample:30: WARNING: Unexpected indentation.
/var/jenkins_home/workspace/yt_docs_new/yt/loaders.py:docstring of yt.loaders.load_sample:33: WARNING: Block quote ends without a blank line; unexpected unindent.
/var/jenkins_home/workspace/yt_docs_new/doc/source/examining/loading_data.rst:4: WARNING: Duplicate explicit target name: "adaptahop".
/var/jenkins_home/workspace/yt_docs_new/doc/source/reference/changelog.rst:45: WARNING: Unexpected indentation.
/var/jenkins_home/workspace/yt_docs_new/doc/source/reference/changelog.rst:46: WARNING: Block quote ends without a blank line; unexpected unindent.
/var/jenkins_home/workspace/yt_docs_new/doc/source/reference/changelog.rst:: WARNING: Anonymous hyperlink mismatch: 1 references but 0 targets.
See "backrefs" attribute for IDs.
/var/jenkins_home/workspace/yt_docs_new/doc/source/visualizing/callbacks.rst:32: WARNING: Mismatch: both interpreted text role prefix and reference suffix.
/var/jenkins_home/workspace/yt_docs_new/doc/source/visualizing/callbacks.rst:159: WARNING: Mismatch: both interpreted text role prefix and reference suffix.
/var/jenkins_home/workspace/yt_docs_new/doc/source/developing/creating_derived_fields.rst:71: WARNING: Mismatch: both interpreted text role prefix and reference suffix.
/var/jenkins_home/workspace/yt_docs_new/doc/source/developing/creating_derived_fields.rst:71: WARNING: Mismatch: both interpreted text role prefix and reference suffix.
```

File: doc/source/reference/api/api.rst [hunk 2]
```diff
@@ -157,27 +157,16 @@ These objects generate an "index" into multiresolution data.
...
-   ~yt.units.yt_array.YTQuantity
-   ~yt.units.yt_array.uconcatenate
-   ~yt.units.yt_array.uintersect1d
-   ~yt.units.yt_array.uunion1d
-   ~yt.units.yt_array.unorm
-   ~yt.units.yt_array.udot
-   ~yt.units.yt_array.uvstack
-   ~yt.units.yt_array.uhstack
-   ~yt.units.yt_array.ustack
-   ~yt.units.yt_array.display_ytarray
```
Comment (matthewturk) [inline doc/source/reference/api/api.rst:180]: Isn't `display_ytarray` still around somewhere?

We may actually want to use intersphinx to point to the unyt definitions of those of these that still exist.  I'm thinking explicitly of the uconcat, uhstack, etc, which are necessary for combining unyt arrays.

File: doc/source/reference/index.rst [hunk 3]
```diff
@@ -13,7 +13,9 @@ code.
    code_support
    command-line
    api/api
+   api/modules
```
Comment (matthewturk) [inline doc/source/reference/index.rst:16]: I have a suspicion this was a deliberate exclusion, but I can't back that up.

**Review by matthewturk — APPROVED** [review]

File: doc/source/reference/index.rst (hunk 3 above)
Comment (neutrinoceros) [reply doc/source/reference/index.rst:16]: the file is generated, it's not clear to me why we would create it but not include it in the index

File: doc/source/reference/api/api.rst (hunk 2 above)
Comment (neutrinoceros) [reply doc/source/reference/api/api.rst:180]: sounds good, but I'm barely managing to fix simple rst errors, honestly I don't really want to invest time in learning how to use intersphinx for now :/

Comment (brittonsmith) [reply doc/source/reference/api/api.rst:180]: I don't mind taking this on.

File: doc/source/analyzing/saving_data.rst [hunk 4]
```diff
@@ -27,7 +27,7 @@ Geometric Data Containers
 -------------------------
 
 Data from geometric data containers can be saved with the
-:func:`~yt.data_objects.data_containers.save_as_dataset`` function.
+:func:`~yt.data_objects.data_containers.save_as_dataset` function.
```
Comment (brittonsmith) [inline doc/source/analyzing/saving_data.rst:30]:
```suggestion
:func:`~yt.data_objects.data_containers.YTDataContainer.save_as_dataset` function.
```

File: doc/source/analyzing/domain_analysis/index.rst [hunk 5]
```diff
@@ -62,9 +63,7 @@ not kept up to date as yt evolved. Tools in here are looking for
 a new owner and a new home. If you find something in here that
 you'd like to bring back to life, either by adding it to
 :ref:`yt-astro` or as part of your own package, you are welcome
-to it! If you'd like any help, let us know! See
-:ref:`yt_attic:attic_modules` for a list of inventory of the
-attic.
+to it! If you'd like any help, let us know!
```
Comment (brittonsmith) [inline doc/source/analyzing/domain_analysis/index.rst:66]:
```suggestion
to it! If you'd like any help, let us know! If you'd like any help, let us know! See
:ref:`yt_attic:attic-modules` for a list of inventory of the
attic.
```

File: doc/source/examining/loading_data.rst [hunk 6]
```diff
@@ -2171,7 +2171,7 @@ containers.  See :ref:`halo_containers` for more information.
 
 If you have access to both the halo catalog and the simulation snapshot from
 the same redshift, additional analysis can be performed for each halo using
-:ref:`halo_catalog`.  The resulting product can be reloaded in a similar manner
+:ref:`halo-catalog`.  The resulting product can be reloaded in a similar manner
```
Comment (brittonsmith) [inline doc/source/examining/loading_data.rst:2174]: This should point to a yt-astro ref.
```suggestion
:ref:`halo-analysis`.  The resulting product can be reloaded in a similar manner
```

File: doc/source/examining/loading_data.rst [hunk 7]
```diff
@@ -2438,12 +2438,11 @@ information.  At this time, halo member particles cannot be loaded.
    # The halo mass
    print(ad["FOF", "particle_mass"])
 
-.. _halocatalog:
+.. _halo-catalog:
```
Comment (brittonsmith) [inline doc/source/examining/loading_data.rst:2441]: This is referred to in yt-astro docs.
```suggestion
.. _halocatalog:
```

File: doc/source/examining/loading_data.rst [hunk 8]
```diff
@@ -2438,12 +2438,11 @@ information.  At this time, halo member particles cannot be loaded.
...
 
 YTHaloCatalog
 ^^^^^^^^^^^^^
 
-These are catalogs produced by the analysis discussed in :ref:`halo_catalog`.
```
Comment (brittonsmith) [inline doc/source/examining/loading_data.rst:2446]: Change ref to halo-analysis

File: doc/source/reference/api/api.rst (hunk 2 above)
Comment (neutrinoceros) [reply doc/source/reference/api/api.rst:180]: thanks !

File: doc/source/examining/loading_data.rst (hunk 8 above)
Comment (neutrinoceros) [reply doc/source/examining/loading_data.rst:2446]: you got it !

File: doc/source/examining/loading_data.rst (hunk 7 above)
Comment (neutrinoceros) [reply doc/source/examining/loading_data.rst:2441]: fair enough !

File: doc/source/analyzing/domain_analysis/index.rst (hunk 5 above)
Comment (neutrinoceros) [reply doc/source/analyzing/domain_analysis/index.rst:66]: I don't get this suggestion. (note that your original suggestions duplicated "If you'd like any help, let us know!", so I edited this out)
Neither references seem to be defined. Am I missing something ?

File: doc/source/analyzing/saving_data.rst (hunk 4 above)
Comment (neutrinoceros) [reply doc/source/analyzing/saving_data.rst:30]: These are two different existing functions, the first one is correct

```
$ wxc yt.data_objects.data_containers.save_as_dataset
/Users/robcleme/dev/yt-project/yt/yt/frontends/ytdata/utilities.py:6
```

File: doc/source/analyzing/domain_analysis/index.rst (hunk 5 above)
Comment (brittonsmith) [reply doc/source/analyzing/domain_analysis/index.rst:66]: The code suggestion doesn't support suggestions for removed lines. All I wanted to point out was that the reference could be renamed and the original text replaced. I had it wrong, though. The reference should just be "attic-modules". It is an intersphinx reference to the `yt_attic` docs, but I forgot it does not need to have `yt_attic:` prepended.

**Review by brittonsmith — APPROVED** [review]

**Pull request merged**